Add change indicator

Eirik0 opened the pull request:
This adds to the json returned when calling z_listreceivedbyaddress and z_listuspent an additional field entitiled 'change' which will be either true or false depending on whether or not the note was change from another transaction.

bitcartel commented:
Hi @Eirik0 can open an issue and link to this PR?  Folk can then discuss on the ticket whether or not there are any privacy concerns from adding a change indicator.  Thanks.

Eirik0 commented:
This is related to issue #2935

Eirik0 commented:
Note: If one sends money from a zaddr to itself, this will count as "change".

str4d commented on qa/rpc-tests/wallet_changeindicator.py:
```diff
@@ -0,0 +1,78 @@
+        # Check zaddr1 received
+        sortedreceived1 = sorted(self.nodes[1].z_listreceivedbyaddress(zaddr1, 0), key = lambda received: received['amount'])
+        assert_equal(2, len(sortedreceived1), "zaddr1 Should have received 2 notes")
+        assert_equal(Decimal('0.4'), sortedreceived1[0]['amount'].quantize(Decimal('1.0')))
```
`.quantize(Decimal('1.0'))` should not be necessary, as the output amounts will already be Decimals, and you are sending the transactions with zero fee. Did you have issues without it?

str4d commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -355,9 +355,18 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
+        raise AssertionError(message)
+        
+def assert_false(condition, message = ""):
+    assert_true(not condition, message)
```
This should ideally be in a separate preceding commit.

Eirik0 replied:
Moved this to a preceding commit.

str4d commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -355,9 +355,18 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
+def assert_equal(expected, actual, message = ""):
+    if expected != actual:
+        if message:
+            message = "%s " % message 
```
This looks like it will result in messages like `Should have received 1 note expected: <X> but was: <Y>`. Add a semicolon before the space.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2552,6 +2553,7 @@ UniValue z_listunspent(const UniValue& params, bool fHelp)
             obj.push_back(Pair("amount", ValueFromAmount(CAmount(entry.plaintext.value))));
             std::string data(entry.plaintext.memo.begin(), entry.plaintext.memo.end());
             obj.push_back(Pair("memo", HexStr(data)));
+            obj.push_back(Pair("change", pwalletMain->IsNoteChange(nullifierSet, entry.address, entry.jsop)));
```
Document this in the help text.

Eirik0 replied:
Done

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3233,25 +3235,26 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
         // (txid, jsindex, jsoutindex) is needed to globally identify a note
         obj.push_back(Pair("jsindex", entry.jsop.js));
         obj.push_back(Pair("jsoutindex", entry.jsop.n));
+        obj.push_back(Pair("change", pwalletMain->IsNoteChange(nullifierSet, entry.address, entry.jsop)));
```
Document this in the help text.

Eirik0 replied:
Done

daira commented on src/wallet/wallet.cpp:
```diff
@@ -421,6 +421,25 @@ void CWallet::SetBestChain(const CBlockLocator& loc)
     SetBestChainINTERNAL(walletdb, loc);
 }
 
+std::set<std::pair<libzcash::PaymentAddress, uint256>> CWallet::GetNullifierSet(const std::set<libzcash::PaymentAddress> & addresses)
```
``GetNullifierRelation``?

Eirik0 commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -355,9 +355,18 @@ def random_transaction(nodes, amount, min_fee, fee_increment, fee_variants):
+        raise AssertionError(message)
+        
+def assert_false(condition, message = ""):
+    assert_true(not condition, message)
```
This has been split out.

Eirik0 commented:
Rebased on to master to resolve merge conflicts.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2461,14 +2461,15 @@ UniValue z_listunspent(const UniValue& params, bool fHelp)
             "\nResult\n"
             "[                             (array of json object)\n"
             "  {\n"
-            "    \"txid\" : \"txid\",          (string) the transaction id \n"
```
Don't remove these spaces; they are necessary for the text to line up in the help output (because the `\"` become `"`).

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2461,14 +2461,15 @@ UniValue z_listunspent(const UniValue& params, bool fHelp)
             "    \"jsoutindex\" : n          (numeric) the output index of the joinsplit\n"
             "    \"confirmations\" : n       (numeric) the number of confirmations\n"
             "    \"spendable\" : true|false  (boolean) true if note can be spent by wallet, false if note has zero confirmations, false if address is watchonly\n"
-            "    \"address\" : \"address\",    (string) the shielded address\n"
```
Ditto.

str4d commented on src/wallet/wallet.cpp:
```diff
@@ -421,6 +421,34 @@ void CWallet::SetBestChain(const CBlockLocator& loc)
     SetBestChainINTERNAL(walletdb, loc);
 }
 
+std::set<std::pair<libzcash::PaymentAddress, uint256>> CWallet::GetNullifiersForAddress(const std::set<libzcash::PaymentAddress> & addresses)
```
`GetNullifiersForAddresses()`

bitcartel replied:
Good idea to change this.

str4d commented on src/wallet/wallet.cpp:
```diff
@@ -421,6 +421,34 @@ void CWallet::SetBestChain(const CBlockLocator& loc)
+    for (const auto & txPair : mapWallet)
+        for (const auto & noteDataPair : txPair.second.mapNoteData)
+            if (noteDataPair.second.nullifier && addresses.count(noteDataPair.second.address))
+                nullifierSet.insert(std::make_pair(noteDataPair.second.address, noteDataPair.second.nullifier.get()));
```
I generally prefer that completely new code adds braces, but non-blocking because it's relatively clear in these instances. Blocking if another reviewer wants it as well.

bitcartel replied:
Yes, please add braces.

Eirik0 commented:
@str4d Addressed your comments and force pushed.

Eirik0 commented:
Worked with @bitcartel on the non-determinism. It seems like this was happening because we were not syncing the nodes before generating new blocks. I was able to reproduce this fairly consistently, but after adding the additional ```sync_all()``` I was no longer able to recreate this.

str4d commented:
The new RPC test timed out on every supported builder.

bitcartel commented:
Strange.  I've run the test locally without any problem.  @Eirik0 Maybe examine the helper function and see if waiting on the async z_sendmany is being executed the same as other qa tests?

str4d commented:
Hmm... The new `wallet_changeindicators` RPC test timed out on one builder, but the `wallet_nullifiers` RPC test failed on the rest.

@zkbot retry

str4d commented:
Same result. There's something strange going on...

Eirik0 commented:
@bitcartel Does that error just pop up intermittently? I've run the test several times and not seen that. I'll look in to it...

Eirik0 commented:
@bitcartel I'm wondering if this is maybe not a timeout thing. I did a bit of digging and it looks like this change has broken the wallet_nullifiers.py test.

Eirik0 commented:
This should now be fixed.

Eirik0 commented:
I think that the timeout might have had to do with trying to initialize a clean chain, so I removed that setup as other similar tests did not seem to be doing that. I also undid the refactoring and changed how we are checking that the lists are equal as per @bitcartel's comment. Force pushed these changes.

Eirik0 commented:
@zkbot try

Eirik0 commented:
@zkbot try

bitcartel commented:
@zkbot r+
